File: app/src/main/java/com/maxwellscore/sparkone/domain/WeatherInteractor.java
package com.maxwellscore.sparkone.domain;

import com.maxwellscore.sparkone.data.WeatherRepository;
import com.maxwellscore.sparkone.domain.entities.Weather;

import io.reactivex.rxjava3.core.Observable;

/**
 * Является представителем слоя UseCase(Domain) и он и модели(дата классы) описывают бизнес логику приложения или фичи
 * В данном случае бизнес-логика простая: Отдать пользователю информацию о погоде
 */
public class WeatherInteractor {

    private WeatherRepository repository = new WeatherRepository();
    private WeatherDataMapper mapper = new WeatherDataMapper();

    public Observable<Weather> getWeather() {
        return repository.getWeatherEasy()
                .map(weatherData -> mapper.dataToDomain(weatherData));
    }
}

File: kotlinexample/src/main/java/com/maxwellscore/kotlinexample/domain/WeatherInteractor.kt
package com.maxwellscore.kotlinexample.domain

import com.maxwellscore.kotlinexample.data.WeatherRepository
import com.maxwellscore.kotlinexample.domain.entities.Weather
import io.reactivex.rxjava3.core.Observable
import javax.inject.Inject

class WeatherInteractor @Inject constructor() {

    private val repository: WeatherRepository = WeatherRepository()
    private val mapper: WeatherDataMapper = WeatherDataMapper()

    fun getWeather(): Observable<Weather> = repository.getWeather().map { mapper.dataToDomain(it) }

    fun postLocation(latitude: Double, longitude: Double) = repository.postLocation(latitude, longitude)
}